Add error checks for bad configuration in `PathFollow2D/3D` `set_progress_ratio`

gturri opened the pull request:
When a `PathFollow2D` is badly configured it's possible to have code that a call to `get_progress_ratio` just after a `set_progress_ratio`, does not return the value just set, which may be confusing for developer (ie that

      myPathFollow2D.set_progress_ratio(0.5)
      myPathFollow2D.get_progress_ratio()

does not return `0.5`)

This commit makes sure the developer has a useful error messages in such case, to make it easier to troubleshoot it.

<!--
Please target the `master` branch in priority.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.

To speed up the contribution process and avoid CI errors, please set up pre-commit hooks locally:
https://docs.godotengine.org/en/latest/contributing/development/code_style_guidelines.html
-->

gturri committed:
Add error checks for bad configuration in `PathFollow2D/3D` `set_progress_ratio`

When a PathFollow is badly configured it's possible to have code that
call get_progress_ratio just after set_progress_ratio does not return
the value just set, which may be confusing for developer (ie that
    myPathFollow2D.set_progress_ratio(0.5)
    myPathFollow2D.get_progress_ratio()
does not return 0.5)

This commit makes ensures the developer has useful error messages in
such case, to make it easier to troubleshot it.

AThousandShips commented on scene/2d/path_2d.cpp:
```diff
@@ -378,9 +378,10 @@ real_t PathFollow2D::get_progress() const {
-	if (path && path->get_curve().is_valid() && path->get_curve()->get_baked_length()) {
-		set_progress(p_ratio * path->get_curve()->get_baked_length());
-	}
+	ERR_FAIL_COND_MSG(!path, "Can't set progress ratio on a PathFollow2D that is not part of the scene tree");
```
```suggestion
	ERR_FAIL_NULL_MSG(path, "Can't set progress ratio on a PathFollow2D that is not part of the scene tree.");
```
Dedicated check, also formatting

gturri replied:
thanks for this feedback.

I was not aware of `ERR_FAIL_NULL_MSG`; it indeed makes more sense to use it 👍 

About this:
> it should say that it's not the child of a path instead I'd say

I'm afraid it's a bit more complex than that, because `PathFollow2D->path` is set only when this node enters the scene tree (see https://github.com/godotengine/godot/blob/master/scene/2d/path_2d.cpp#L259 ).
Which means that if I set a `PathFollow2D` as a child of a `Path2D` but then forget to add the `Path2D` in my scene tree, I'll face this behavior were set is silently doing nothing.
But you're correct: `path` may be null if the parent is not a `Path2D`, so I guess the message could be reworded as: 

> Can only set progress ratio on a PathFollow2D that is the child of a Path2D which is itself part of the scene tree.

Does that make more sense like this?

AThousandShips replied:
That sounds good

AThousandShips commented on scene/2d/path_2d.cpp:
```diff
@@ -378,9 +378,10 @@ real_t PathFollow2D::get_progress() const {
-		set_progress(p_ratio * path->get_curve()->get_baked_length());
-	}
+	ERR_FAIL_COND_MSG(!path, "Can't set progress ratio on a PathFollow2D that is not part of the scene tree");
+	ERR_FAIL_COND_MSG(!path->get_curve().is_valid(), "Can't set progress ratio on a PathFollow2D that does not have a Curve");
```
```suggestion
	ERR_FAIL_COND_MSG(path->get_curve().is_null(), "Can't set progress ratio on a PathFollow2D that does not have a Curve.");
```

AThousandShips commented on scene/2d/path_2d.cpp:
```diff
@@ -378,9 +378,10 @@ real_t PathFollow2D::get_progress() const {
-	}
+	ERR_FAIL_COND_MSG(!path, "Can't set progress ratio on a PathFollow2D that is not part of the scene tree");
+	ERR_FAIL_COND_MSG(!path->get_curve().is_valid(), "Can't set progress ratio on a PathFollow2D that does not have a Curve");
+	ERR_FAIL_COND_MSG(!path->get_curve()->get_baked_length(), "Can't set progress ratio on a PathFollow2D that has a 0 length curve");
```
```suggestion
	ERR_FAIL_COND_MSG(!path->get_curve()->get_baked_length(), "Can't set progress ratio on a PathFollow2D that has a 0 length curve.");
```

AThousandShips commented:
This introduces errors where none were before, this is not necessarily appropriate

gturri commented:
Thanks for this feedback.

About

> This introduces errors where none were before

Yes, correct. And it turns out that's the point.

Some context (that I could have shared earlier, sorry for that): I had code like this

```
var path = Path2D.new()
var curve = Curve2D.new()
curve.add_point(Vector2.ZERO)
curve.add_point(Vector2(1, 1))
path.curve = curve
var pathFollow2D = PathFollow2D.new()
path.add_child(pathFollow2D)

pathFollow2D.progress_ratio = 0.42
print("progress_ratio: " + str(pathFollow2D.progress_ratio))
```

and it took me dozen of minutes to understand why the value printed was `0` instead of the `0.42` I set just a line above.
I eventually understood it because I fortunately did some C++ in a previous life and could hence read the godot code. So I could understand that my issue was that `path` was null, and understand that the way to have it non-null is to add the parent `Path2D` to the tree scene.
I guess that those error messages not only could have saved me a bit of time, I also guess that some people less fluent in C++ could appreciate some error message to have clues about what they are doing wrong.

About 

> makes the code significantly different from the 3D case

thanks for drawing my attention on it. It indeed would make sense to have in the 3D case as well.



--
So I guess my question now is: does it make sense that I apply the suggested changes (namely:

- fix formatting
- use `ERR_FAIL_NULL_MSG`
- make it explicit in the first error message that the parent should be a `Path2D`
- apply the same change in the 3D case
)

or do you feel it's a change that you would not want merged anyway?

AThousandShips commented:
I think this might be better to document instead, but I'm fine with either way, but:
* This should be in both 2D and 3D or neither IMO
* It should also be documented in either case

gturri commented:
Thanks for that answer.

Just to make sure I understand correctly: when you say that it should be documented, do you mean that I should do a PR on the documentation repo in order to make it clear on the doc of the `set_progress_ratio` field that it cannot be changed in the cases identified (ie: are we talking of a PR on this: https://github.com/godotengine/godot-docs/blob/4b755c0dac449ec9cd899dbf39d7619fb0edba2e/classes/class_pathfollow2d.rst#L13 ), or do you have some other documentation in mind?

AThousandShips commented:
No it should be here, documentation is generated from `doc/classes/*.xml` files, that should be added in this PR as well, or as an alternative PR just documenting this limitation

gturri commented:
Thanks for this pointer.

I should be able to take the time to do this change today or tomorrow 👍

gturri commented:
I just updated that pull request accordingly, with both the changes in the code (fixes for `PathFollow2D` + applied the changes for `PathFollow3D`) and updated the doc.

Don't hesitate to let me know if there are other things you'd rather I change.

KoBeWi reviewed (approved):
I wonder if `get_progress_ratio()` should also have similar error. Technically this will fail silently too:
```GDScript
myPathFollow2D.progress = 100
print(myPathFollow2D.progress_ratio) # 0
```

gturri commented:
> I wonder if get_progress_ratio() should also have similar error
yes, correct: it has the same behavior. That's why the doc updated by this commit also mentions the getter.

I did not add the `ERR_FAIL_NULL_MSG` & co in the getter because I though that if someone manipulates a `PathFollow2D` when e.g. the `Path2D` is not part of the scene tree, then he or she will already have an error message upon calling the setter, which already makes it easy to understand what is going wrong.
(also, in the [1st comment on this PR](https://github.com/godotengine/godot/pull/94886#issuecomment-2255336934) someone expresses concerns about introducing new errors, so I figured out it would be better to not add more of those).

That being said, if you believe it would be better to emit those error messages in the getter, I can amend this PR.

gturri commented:
Hi there,

I understand that you are all pretty busy with other stuff (like summer holidays, or shipping godot 4.3 (thanks btw), or anything else), so it's cool that this PR is on hold.
On the other hand, if the current status is that something is expected from me in order to make this PR progress, please don't hesitate to let me know :)

timothyqiu commented on doc/classes/PathFollow2D.xml:
```diff
@@ -26,6 +26,7 @@
 		</member>
 		<member name="progress_ratio" type="float" setter="set_progress_ratio" getter="get_progress_ratio" default="0.0">
 			The distance along the path as a number in the range 0.0 (for the first vertex) to 1.0 (for the last). This is just another way of expressing the progress within the path, as the offset supplied is multiplied internally by the path's length.
+			It can be set or get only if the [PathFollow2D] is the child of a [Path2D] which is part of the scene tree, and that this [Path2D] has a [Curve2D] with a non-zero length. Otherwise, trying to set this field will have no effect, and getting this field will return [code]0.0[/code].
```
> trying to set this field will have no effect

It _did_ not work before, but since it will report an error now, I think it can be something like "trying to set this field will print an error".

The same applies to the 3D version.

gturri replied:
Thanks for this feedback.
Indeed, it makes sense. I just updated the PR accordingly.

akien-mga reviewed (approved):
Makes sense to me.

I force pushed an amend of the commit to make it match the PR title, which is more informative.

akien-mga commented:
Thanks! And congrats for your first merged Godot contribution :tada:
